### DISCUSSION TITLE
Xpra clipping optimization

### DISCUSSION THREAD

**Discussion by TheDude529** [Xpra-org/xpra#4387] [General]
Hi,

First, I must say that Xpra is great,  kudos ! 

I wanted to ask about how (and if) Xpra optimize 2 common cases, in which updating the client is rather redundant :

1) I'm more of a low-level guy, specially in linux, and not familiar much w/ linux graphical sw stack . 
I wonder about cases where the actual app draws on rectangles that are practically hidden due to other parts override them, 
Where in the stack Xpra resides ?   
A simple example would be,  gnome-terminal w/ 2 tabs inside the window,  switching between them from time to time . 
Suppose one tab busy w/ constant prints on the terminal, however, switching to the second tab will effectively hide the first tab and it's print .
If Xpra resides on the first graphical layers below the app, it will surely respond and perhaps even try to update the client on the other side of the prints (the draws), after which the client on the other side will discard them as it is a hidden tab,      if Xpra resides purely on the lowest per-window layer (it's a windowed session, not a desktop),   it will get only updates of the current visible terminal which is much more optimal .   

2) If there are several windows shared between a server and a client, is there an optimization in which the client updates the server when a window is being minimized / fully hidden,  so that the server won't bloat the client w/ updates for a window that is not visible at all,  and when the user restores the window, the client will ask the server for the last img ? 

I know this means that there will be a delay during window restore, but otherwise, if working with several windows, the client keeps update the windows as far as I understand (??), and for a laptop client it can get pretty busy / hot .
(or is there a tweak / config hack that might make this desired behavior ?)

Thanks a lot anyway !

Dude

**Comment by totaam** [comment]
> Where in the stack Xpra resides ?

It only communicates with the X11 server directly, not the applications themselves. It knows nothing about tabs, each window is just a big rectangular area with pixels as far as it is concerned.
The X11 server tells xpra when the areas of each window that applications have (re)painted.

> If there are several windows shared between a server and a client...

xpra will prioritize the window which has focus, full-screen windows and maximized windows.
If a window is minimized or unmapped, it is also unmapped server side and there are no pixels to send since they do not exist.
It will also drastically slow down the refresh rate when a window is not on a visible workspace, when the client is suspending, etc
The refresh rate should also always be lower than the client's monitor refresh rate.
